File: Day048_2025/Q048_Leetcode.java
class Q048_Leetcode {
    private int findSequences(int[] charCount) {
        int totalCount = 0;
        
        for (int pos = 0; pos < 26; pos++) {
            if (charCount[pos] == 0) {
                continue;
            }
            totalCount++; // Counting the current sequence
            
            charCount[pos]--; // Use one occurrence of the character
            totalCount += findSequences(charCount); // Recursive call
            charCount[pos]++; // Backtrack
        }
        
        return totalCount;
    }

    public int numTilePossibilities(String tiles) {
        int[] charCount = new int[26];
        
        for (char c : tiles.toCharArray()) {
            charCount[c - 'A']++;
        }
        
        return findSequences(charCount);
    }
}

File: Day048_2025/Q048_Leetcode.py
class Solution:
    def findSequences(self, charCount):
        totalCount = 0

        for pos in range(26):
            if charCount[pos] == 0:
                continue
            totalCount += 1  # Counting the current sequence

            charCount[pos] -= 1  # Use one occurrence of the character
            totalCount += self.findSequences(charCount)  # Recursive call
            charCount[pos] += 1  # Backtrack

        return totalCount

    def numTilePossibilities(self, tiles: str) -> int:
        charCount = [0] * 26
        
        for c in tiles:
            charCount[ord(c) - ord('A')] += 1
        
        return self.findSequences(charCount)
